geode-core/src/main/java/org/apache/geode/internal/cache/TXState.java

```diff
@@ -1000,11 +1004,17 @@ public void beforeCompletion() throws SynchronizationCommitConflictException {
     if (this.closed) {
       throw new TXManagerCancelledException();
     }
-    this.proxy.getTxMgr().setTXState(null);
+    if (beforeCompletionCalled) {
+      // do not re-execute beforeCompletion again
```

Comment:
this code does not look thread safe. "beforeCompletionCalled" is volatile which implies it could be used by more than one thread. Do we really have cases in which multiple threads call beforeCompletion? If so then should "beforeCompletion" be changed to synchronized method? Then this code would be correct and the field would not need to be volatile.
You would also want to make "afterCompletion" synchronized which I think makes sense if two different threads could call these two methods. That way if beforeCompletion is currently in progress any call of afterCompletion would wait for the beforeCompletion it finish.

Reply:
It is possible for multiple threads to call beforeCompletion in failover case. However, it is thread safe as only one thread could hold the lock for the TXStateProxyImpl when TXManagerImpl.masqueradeAs() is invoked.

Reply:
Also afterCompletion will not be invoked until beforeCompletion is executed on the tx host (TXState) if it is a commit. It also needs to hold the lock in TXStateProxyImpl, before it can invoke afterCompletion. I think it is safe.

Reply:
I think it would probably be best to make these methods synchronized.
It sounds like we never expect two threads to concurrently contend for this synchronization so it not cost in performance (compared to the current volatile read/write). 
With the synchronized you know that only one is in progress at a time and can tell by just looking at the code in this class that it is thread safe.

Reply:
Made the suggested change.